```bash
routes:

  - name: chat
    route_type: llm/v1/chat
    model:
      provider: openai
      name: gpt-4
      config:
        openai_api_base: https://api.openai.com/v1
        openai_api_key: $OPENAI_API_KEY

  - name: completions-openai
    route_type: llm/v1/completions
    model:
      provider: openai
      name: gpt-4
      config:
        openai_api_base: https://api.openai.com/v1
        openai_api_key: $OPENAI_API_KEY

  - name: completions-claude
    route_type: llm/v1/completions
    model:
      provider: anthropic
      name: claude-1.3-100k
      config:
        anthropic_api_base: https://api.anthropic.com/v1
        anthropic_api_key: $ANTHROPIC_API_KEY

  - name: completions-cohere
    route_type: llm/v1/completions
    model:
      provider: cohere
      name: command
      config:
        cohere_api_key: $COHERE_API_KEY

  - name: chat-mosaic
    route_type: llm/v1/chat
    model:
      provider: mosaicml
      name: llama2-70b-chat
      config:
        mosaicml_api_key: $MOSAICML_API_KEY

  - name: completions-mosaic
    route_type: llm/v1/completions
    model:
      provider: mosaicml
      name: llama2-70b-chat
      config:
        mosaicml_api_key: $MOSAICML_API_KEY

  - name: embeddings-mosaic
    route_type: llm/v1/embeddings
    model:
      provider: mosaicml
      name: instructor-xl
      config:
        mosaicml_api_key: $MOSAICML_API_KEY

  - name: embeddings-oss
    route_type: llm/v1/embeddings
    model:
      provider: mlflow-model-serving
      name: mini-lm-l6-v2
      config:
        model_server_url: http://127.0.0.1:9010

  - name: completions-oss
    route_type: llm/v1/completions
    model:
      provider: mlflow-model-serving
      name: emo-bert
      config:
        model_server_url: http://127.0.0.1:9020

  - name: bad-mosaic
    route_type: llm/v1/chat
    model:
      provider: mosaicml
      name: invalid
      config:
        mosaicml_api_key: $MOSAICML_API_KEY
```

Last entry is a bad config (chat route with invalid model name)

Auto-updating config with running server:

```shell
➜ mlflow gateway start --config-path ~/gateway/config.yaml --port 9100 --host localhost
[2023-08-29 19:09:18 -0400] [42656] [INFO] Starting gunicorn 20.1.0
[2023-08-29 19:09:18 -0400] [42656] [INFO] Listening at: http://127.0.0.1:9100 (42656)
[2023-08-29 19:09:18 -0400] [42656] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2023-08-29 19:09:18 -0400] [42659] [INFO] Booting worker with pid: 42659
[2023-08-29 19:09:18 -0400] [42660] [INFO] Booting worker with pid: 42660
[2023-08-29 19:09:22 -0400] [42660] [INFO] Started server process [42660]
[2023-08-29 19:09:22 -0400] [42659] [INFO] Started server process [42659]
[2023-08-29 19:09:22 -0400] [42660] [INFO] Waiting for application startup.
[2023-08-29 19:09:22 -0400] [42659] [INFO] Waiting for application startup.
[2023-08-29 19:09:22 -0400] [42659] [INFO] Application startup complete.
[2023-08-29 19:09:22 -0400] [42660] [INFO] Application startup complete.
2023/08/30 14:00:49 INFO mlflow.gateway.runner: Configuration updated, reloading workers
[2023-08-30 14:00:49 -0400] [42659] [INFO] Shutting down
[2023-08-30 14:00:49 -0400] [42659] [INFO] Error while closing socket [Errno 9] Bad file descriptor
[2023-08-30 14:00:49 -0400] [42660] [INFO] Shutting down
[2023-08-30 14:00:49 -0400] [42660] [INFO] Error while closing socket [Errno 9] Bad file descriptor
[2023-08-30 14:00:49 -0400] [42659] [INFO] Waiting for application shutdown.
[2023-08-30 14:00:49 -0400] [42659] [INFO] Application shutdown complete.
[2023-08-30 14:00:49 -0400] [42659] [INFO] Finished server process [42659]
[2023-08-30 14:00:49 -0400] [42660] [INFO] Waiting for application shutdown.
[2023-08-30 14:00:49 -0400] [42660] [INFO] Application shutdown complete.
[2023-08-30 14:00:49 -0400] [42659] [INFO] Worker exiting (pid: 42659)
[2023-08-30 14:00:49 -0400] [42660] [INFO] Finished server process [42660]
[2023-08-30 14:00:49 -0400] [42660] [INFO] Worker exiting (pid: 42660)
[2023-08-30 14:00:50 -0400] [47231] [INFO] Booting worker with pid: 47231
[2023-08-30 14:00:50 -0400] [47232] [INFO] Booting worker with pid: 47232
[2023-08-30 14:00:51 -0400] [47232] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/uvicorn/workers.py", line 66, in init_process
    super(UvicornWorker, self).init_process()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/util.py", line 412, in import_app
    app = app(*args, **kwargs)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/app.py", line 226, in create_app_from_env
    return create_app_from_path(config_path)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/app.py", line 217, in create_app_from_path
    config = _load_route_config(config_path)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/config.py", line 366, in _load_route_config
    return GatewayConfig(**configuration)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 404, in __init__
    values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 1040, in validate_model
    v_, errors_ = field.validate(value, values, loc=field.alias, cls=cls_)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 734, in validate
    v, errors = self._validate_sequence_like(v, values, loc, cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 767, in _validate_sequence_like
    r, ee = self._validate_singleton(v_, values, v_loc, cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 899, in _validate_singleton
    value, error = field.validate(v, values, loc=loc, cls=cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 723, in validate
    v, errors = self._validate_singleton(v, values, loc, cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 906, in _validate_singleton
    return self._apply_validators(v, values, loc, cls, self.validators)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 913, in _apply_validators
    v = validator(cls, v, values, self, self.model_config)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/class_validators.py", line 310, in <lambda>
    return lambda cls, v, values, field, config: validator(v)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 735, in validate
    return cls(**value)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 404, in __init__
    values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 1066, in validate_model
    values = validator(cls_, values)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/config.py", line 295, in validate_route_type_and_model_name
    raise MlflowException.invalid_parameter_value(
mlflow.exceptions.MlflowException: An invalid model has been specified for the chat route. 'invalid'. Ensure the model selected starts with one of: ['llama2']
[2023-08-30 14:00:51 -0400] [47231] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 589, in spawn_worker
    worker.init_process()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/uvicorn/workers.py", line 66, in init_process
    super(UvicornWorker, self).init_process()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/workers/base.py", line 134, in init_process
    self.load_wsgi()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/workers/base.py", line 146, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/util.py", line 412, in import_app
    app = app(*args, **kwargs)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/app.py", line 226, in create_app_from_env
    return create_app_from_path(config_path)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/app.py", line 217, in create_app_from_path
    config = _load_route_config(config_path)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/config.py", line 366, in _load_route_config
    return GatewayConfig(**configuration)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 404, in __init__
    values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 1040, in validate_model
    v_, errors_ = field.validate(value, values, loc=field.alias, cls=cls_)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 734, in validate
    v, errors = self._validate_sequence_like(v, values, loc, cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 767, in _validate_sequence_like
    r, ee = self._validate_singleton(v_, values, v_loc, cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 899, in _validate_singleton
    value, error = field.validate(v, values, loc=loc, cls=cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 723, in validate
    v, errors = self._validate_singleton(v, values, loc, cls)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 906, in _validate_singleton
    return self._apply_validators(v, values, loc, cls, self.validators)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/fields.py", line 913, in _apply_validators
    v = validator(cls, v, values, self, self.model_config)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/class_validators.py", line 310, in <lambda>
    return lambda cls, v, values, field, config: validator(v)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 735, in validate
    return cls(**value)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 404, in __init__
    values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/pydantic/main.py", line 1066, in validate_model
    values = validator(cls_, values)
  File "/Users/benjamin.wilson/repos/mlflow-fork/mlflow/mlflow/gateway/config.py", line 295, in validate_route_type_and_model_name
    raise MlflowException.invalid_parameter_value(
mlflow.exceptions.MlflowException: An invalid model has been specified for the chat route. 'invalid'. Ensure the model selected starts with one of: ['llama2']
[2023-08-30 14:00:51 -0400] [47232] [INFO] Worker exiting (pid: 47232)
[2023-08-30 14:00:51 -0400] [47231] [INFO] Worker exiting (pid: 47231)
Traceback (most recent call last):
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 209, in run
    self.sleep()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 357, in sleep
    ready = select.select([self.PIPE[0]], [], [], 1.0)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/__main__.py", line 7, in <module>
    run()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 67, in run
    WSGIApplication("%(prog)s [OPTIONS] [APP_MODULE]").run()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/base.py", line 231, in run
    super().run()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/app/base.py", line 72, in run
    Arbiter(self).run()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 229, in run
    self.halt(reason=inst.reason, exit_status=inst.exit_status)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 342, in halt
    self.stop()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 393, in stop
    time.sleep(0.1)
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
    self.reap_workers()
  File "/Users/benjamin.wilson/miniconda3/envs/mlflow-dev-env/lib/python3.8/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
```

Removed the invalid route entry and...

```shell
➜ mlflow gateway start --config-path ~/gateway/config.yaml --port 9100 --host localhost
[2023-08-30 14:06:00 -0400] [57189] [INFO] Starting gunicorn 20.1.0
[2023-08-30 14:06:00 -0400] [57189] [INFO] Listening at: http://127.0.0.1:9100 (57189)
[2023-08-30 14:06:00 -0400] [57189] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2023-08-30 14:06:00 -0400] [57191] [INFO] Booting worker with pid: 57191
[2023-08-30 14:06:00 -0400] [57192] [INFO] Booting worker with pid: 57192
[2023-08-30 14:06:02 -0400] [57191] [INFO] Started server process [57191]
[2023-08-30 14:06:02 -0400] [57191] [INFO] Waiting for application startup.
[2023-08-30 14:06:02 -0400] [57191] [INFO] Application startup complete.
[2023-08-30 14:06:02 -0400] [57192] [INFO] Started server process [57192]
[2023-08-30 14:06:02 -0400] [57192] [INFO] Waiting for application startup.
[2023-08-30 14:06:02 -0400] [57192] [INFO] Application startup complete.
```
Good to go!

--- a/mlflow/gateway/config.py
+++ b/mlflow/gateway/config.py
@@ -15,7 +15,7 @@ from pydantic.json import pydantic_encoder
 from mlflow.exceptions import MlflowException
 from mlflow.gateway.base_models import ConfigModel, ResponseModel
 from mlflow.gateway.constants import (
-    MLFLOW_AI_GATEWAY_MOSAIC_CHAT_SUPPORTED_MODEL_PREFIXES,
+    MLFLOW_AI_GATEWAY_MOSAICML_CHAT_SUPPORTED_MODEL_PREFIXES,
     MLFLOW_GATEWAY_ROUTE_BASE,
     MLFLOW_QUERY_SUFFIX,
 )
@@ -295,7 +295,7 @@ class RouteConfig(ConfigModel):
             raise MlflowException.invalid_parameter_value(
                 f"An invalid model has been specified for the chat route. '{model.name}'. "
                 f"Ensure the model selected starts with one of: "
-                f"{MLFLOW_AI_GATEWAY_MOSAIC_CHAT_SUPPORTED_MODEL_PREFIXES}"
+                f"{MLFLOW_AI_GATEWAY_MOSAICML_CHAT_SUPPORTED_MODEL_PREFIXES}"
             )
         return values
 

--- a/mlflow/gateway/constants.py
+++ b/mlflow/gateway/constants.py
@@ -36,4 +36,4 @@ MLFLOW_SERVING_RESPONSE_KEY = "predictions"
 # These validated names are used for the MosaicML provider due to the need to perform prompt
 # translations prior to sending a request payload to their chat endpoints.
 # to reduce the need to case-match, supported model prefixes are lowercase.
-MLFLOW_AI_GATEWAY_MOSAIC_CHAT_SUPPORTED_MODEL_PREFIXES = ["llama2"]
+MLFLOW_AI_GATEWAY_MOSAICML_CHAT_SUPPORTED_MODEL_PREFIXES = ["llama2"]

--- a/mlflow/gateway/utils.py
+++ b/mlflow/gateway/utils.py
@@ -10,7 +10,7 @@ import psutil
 
 from mlflow.environment_variables import MLFLOW_GATEWAY_URI
 from mlflow.exceptions import MlflowException
-from mlflow.gateway.constants import MLFLOW_AI_GATEWAY_MOSAIC_CHAT_SUPPORTED_MODEL_PREFIXES
+from mlflow.gateway.constants import MLFLOW_AI_GATEWAY_MOSAICML_CHAT_SUPPORTED_MODEL_PREFIXES
 from mlflow.utils.annotations import experimental
 from mlflow.utils.uri import append_to_uri_path
 
@@ -172,5 +172,5 @@ class SearchRoutesToken:
 def is_valid_mosiacml_chat_model(model_name: str) -> bool:
     return any(
         model_name.lower().startswith(supported)
-        for supported in MLFLOW_AI_GATEWAY_MOSAIC_CHAT_SUPPORTED_MODEL_PREFIXES
+        for supported in MLFLOW_AI_GATEWAY_MOSAICML_CHAT_SUPPORTED_MODEL_PREFIXES
     )

